File: content/snippets/cpp/interview_lcci/04/05_legal_binary_search_tree.cpp
#include <iostream>
#include <limits>

struct Node { int v; Node* l; Node* r; Node(int x): v(x), l(nullptr), r(nullptr) {} };

bool isValid(Node* n, long long lo, long long hi) {
    if (!n) return true;
    if (n->v <= lo || n->v >= hi) return false;
    return isValid(n->l, lo, n->v) && isValid(n->r, n->v, hi);
}

int main() {
    Node* ok = new Node(5); ok->l = new Node(3); ok->r = new Node(7);
    Node* bad = new Node(5); bad->l = new Node(3); bad->r = new Node(7); bad->l->r = new Node(6);
    std::cout << std::boolalpha << isValid(ok, std::numeric_limits<long long>::min(), std::numeric_limits<long long>::max()) << '\n';
    std::cout << std::boolalpha << isValid(bad, std::numeric_limits<long long>::min(), std::numeric_limits<long long>::max()) << '\n';
}

File: content/snippets/python/interview_lcci/04/05_legal_binary_search_tree.py
class Node:
    def __init__(self, val: int) -> None:
        self.val = val
        self.left: Node | None = None
        self.right: Node | None = None


def is_valid(node: Node | None, lo: int, hi: int) -> bool:
    if node is None:
        return True
    if not (lo < node.val < hi):
        return False
    return is_valid(node.left, lo, node.val) and is_valid(node.right, node.val, hi)


def main() -> None:
    ok = Node(5); ok.left = Node(3); ok.right = Node(7)
    bad = Node(5); bad.left = Node(3); bad.right = Node(7); bad.left.right = Node(6)
    print(is_valid(ok, -10**18, 10**18))
    print(is_valid(bad, -10**18, 10**18))


if __name__ == "__main__":
    main()
